writers_rollingfilewriter.go

```diff
@@ -328,54 +346,102 @@ func (rw *rollingFileWriter) createFileAndFolderIfNeeded(first bool) error {
 	return nil
 }
 
-func (rw *rollingFileWriter) archiveExplodedLogs(logFilename string, compressionType compressionType) error {
+func (rw *rollingFileWriter) archiveExplodedLogs(logFilename string, compressionType compressionType) (err error) {
 	rollPath := filepath.Join(rw.currentDirPath, logFilename)
-	bts, err := ioutil.ReadFile(rollPath)
+	src, err := os.Open(rollPath)
 	if err != nil {
 		return err
 	}
+	defer src.Close() // Read-only
 
-	entry := make(map[string][]byte)
-	entry[logFilename] = bts
 	archiveFile := path.Clean(rw.archivePath + "/" + compressionType.rollingArchiveTypeName(logFilename, true))
+	dst, err := os.Create(archiveFile)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if cerr := dst.Close(); cerr != nil && err == nil {
+			err = cerr
+		}
+	}()
 
 	// archive entry
-	return compressionType.archive(archiveFile, entry, true)
+	w := compressionType.archiver(dst, true)
+	defer func() {
+		if cerr := w.Close(); cerr != nil && err == nil {
+			err = cerr
+		}
+	}()
+	fi, err := src.Stat()
+	if err != nil {
+		return err
+	}
+	w.NextFile(logFilename, fi)
+	_, err = io.Copy(w, src)
+	return err
 }
 
-func (rw *rollingFileWriter) archiveUnexplodedLogs(compressionType compressionType, rollsToDelete int, history []string) error {
-	var files map[string][]byte
-	// If archive exists
-	_, err := os.Lstat(rw.archivePath)
-	if nil == err {
-		// Extract files and content from it
-		files, err = compressionType.unarchive(rw.archivePath)
+func (rw *rollingFileWriter) archiveUnexplodedLogs(compressionType compressionType, rollsToDelete int, history []string) (err error) {
+	// Buffer to a temporary file
+	dst, err := ioutil.TempFile("", "archived_logs")
+	if err != nil {
+		return err
+	}
+	defer dst.Close()           // Temporary file
+	defer os.Remove(dst.Name()) // Temporary file
+
+	w := compressionType.archiver(dst, false)
+	defer func() {
+		if cerr := w.Close(); cerr != nil && err == nil {
+			err = cerr
+		}
+	}()
+
+	_, err = os.Lstat(rw.archivePath)
+	switch {
```

Comment:
Not major, but pretty sure you can inline assignment to switch if you're only using variables in content:

```go
switch _, err := os.Lstat(rw.archivePath) {
...
}
```

I know `err` is shared with the named return variable, but don't think it matters in this case since all paths in switch return on non-nil error (although if you want to keep it this way for clarity it's fine)